javascript/src/refactor.js

```diff
@@ -0,0 +1,648 @@
+var estraverse = require('estraverse'); // Convenient API for AST traversal                       
+const winston = require('winston'); // logger
+const colors = require('colors');
+
+class RefactorEngine {
+    constructor(ast, 
+                properties, 
+                behaviour, 
+                flagname, 
+                max_cleanup_steps, 
+                print_to_console = false, 
+                keep_comments = false,
+                filename = null) {
+        this.ast = ast;
+        this.properties = properties;
+        this.flagname = flagname;
+        this.max_cleanup_steps = max_cleanup_steps;
+        this.print_to_console = print_to_console;
+        this.keep_comments = keep_comments;
+        this.behaviour = behaviour;
+        this.changed = false;
+        this.filename = filename;
+    }
+        
+    trueLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: true, 
+            raw: 'true'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+    
+    falseLiteral() { 
+        const vanillaLiteral = { // A Boolean Literal is parsed like this by recast
+            type: "Literal", 
+            value: false, 
+            raw: 'false'
+        };
+    
+        // This additional property marks that this literal was introduced by Piranha 
+        vanillaLiteral.createdByPiranha = true;
+    
+        return vanillaLiteral;
+    }
+
+    getMethodHashMap(properties) { 
+        var methodHashMap = new Map();
+    
+        properties["methodProperties"].forEach((item) => {
+            methodHashMap.set(item["methodName"], item);
+        });
+    
+        return methodHashMap;
+    }
+
+    // Verify this is a literal introduced by Piranha. 
+    // We only refactor code associated with literals having this property and leave original code untouched
+    isPiranhaLiteral(node) {
+        return node.type === "Literal" && node.createdByPiranha !== undefined;
```

Comment:
Cool. Wish I could annotate AST nodes as generated by Piranha like this on the Java side 😉